lemminx-liberty/src/main/java/io/openliberty/tools/langserver/lemminx/LibertyDocumentLinkParticipant.java

```diff
@@ -0,0 +1,63 @@
+package io.openliberty.tools.langserver.lemminx;
+
+import java.util.List;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.stream.Collectors;
+
+import org.apache.xerces.impl.XMLEntityManager;
+import org.apache.xerces.util.URI.MalformedURIException;
+import org.eclipse.lemminx.dom.DOMAttr;
+import org.eclipse.lemminx.dom.DOMDocument;
+import org.eclipse.lemminx.dom.DOMNode;
+import org.eclipse.lemminx.services.extensions.IDocumentLinkParticipant;
+import org.eclipse.lemminx.utils.XMLPositionUtility;
+import org.eclipse.lsp4j.DocumentLink;
+import org.eclipse.lsp4j.Range;
+
+import io.openliberty.tools.langserver.lemminx.util.LibertyUtils;
+
+
+public class LibertyDocumentLinkParticipant implements IDocumentLinkParticipant {
+
+    private static final Logger LOGGER = Logger.getLogger(LibertyDocumentLinkParticipant.class.getName());
+
+    @Override
+    public void findDocumentLinks(DOMDocument document, List<DocumentLink> links) {
+
+        if (!LibertyUtils.isConfigXMLFile(document))
+            return;
+
+        List<DOMNode> childNodes = document.getDocumentElement().getChildren();
+
+        // collect all <include> nodes that are children of the document element
+        List<DOMNode> includeDomNodes = childNodes.stream().filter(n -> n.getNodeName().equals("include"))
+                .collect(Collectors.toList());
+
+        for (DOMNode includeNode : includeDomNodes) {
+            DOMAttr includeAttr = includeNode.getAttributeNode("location");
+            
+            String locAttr = includeAttr.getValue();
+            if (!locAttr.endsWith(".xml") || locAttr.startsWith("http") || locAttr.contains("$")) {
+                // ignoring http or $ vars until feature supported
+                continue;
+            }
+            Range linkRange = XMLPositionUtility.selectAttributeValue(includeAttr);
+            try {
+                String location = getResolvedLocation(document.getDocumentURI(), includeAttr.getValue());
```

Comment:
So do we know it is a valid location at this point? Or do we go into the `catch` below if it is not?

Reply:
We go into the `catch` if the URI couldn't be expanded/generated for the link. We don't particularly test for validity -- diagnostics does that

Reply:
When you did the demo, it seemed like the document link only got created when it was an actual valid location. That is why I asked the question. I think your answer is saying it would go into the `catch` if it was not a valid location. Once the document link is created, I understand from your demo that the location could be changed to something invalid, and then we rely on our diagnostics to flag that. Did I get that right?

Reply:
Ah, yes